### ISSUE TITLE
bug: react - maximum depth exceeded from PrivateRoute

### ISSUE THREAD

**Issue by jeffcjohnson** [ionic-team/ionic-framework#21717]
# Bug Report

**Ionic version:**
[ ] **4.x**
[x] **5.x**

**Current behavior:**
Redirecting from a PrivateRoute causes "Maximum update depth exceeded".

**Expected behavior:**
Redirect to, for example, a "/login" page.

I have been trying to use the PrivateRoute pattern described [here](https://reactrouter.com/web/example/auth-workflow) without success.

**Steps to reproduce:**

`ionic start ionic-react-redirect-bug conference --type=react`
`cd ionic-react-redirect-bug`

Add the following to `src/App.tsx`:
```
interface PrivateRouteProps extends RouteProps {
  component: any;
}

const PrivateRoute = ({component: Component, ...rest}: PrivateRouteProps) => {
  const isLoggedIn = false;

  return (
    <Route
      {...rest}
      render={(routeProps) =>
        isLoggedIn ? (
          <Component {...routeProps} />
        ) : (
          <Redirect to={{pathname: "/login", state: {from: routeProps.location}}}/>
        )
      }
    />
  );
}
```

In `src/App.tsx` replace:

`<Route path="/support" component={Support} />`

with:

`<PrivateRoute path="/support" component={Support} />`

Comment out `useEffect` in `src/components/Map.tsx` if it complains about anything (unrelated to this issue).

`ionic serve`

Click the `Support` menu item.

**Other information:**

In the browser console:

```
Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    React 4
    unlisten react-router.js:69
    listener history.js:162
    notifyListeners history.js:180
    notifyListeners history.js:179
    setState history.js:300
    replace history.js:414
    confirmTransitionTo history.js:152
    replace history.js:397
    onUpdate react-router.js:309
    componentDidUpdate react-router.js:189
    React 6
    unstable_runWithPriority scheduler.development.js:659
    React 5
    unstable_runWithPriority scheduler.development.js:659
    React 6
```

**Ionic info:**

```
Ionic:

   Ionic CLI       : 6.10.1 (/usr/local/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 5.2.3

Capacitor:

   Capacitor CLI   : 1.3.0
   @capacitor/core : 1.3.0

Utility:

   cordova-res : not installed
   native-run  : 1.0.0

System:

   NodeJS : v14.4.0 (/usr/local/Cellar/node/14.4.0/bin/node)
   npm    : 6.14.6
   OS     : macOS Catalina
```

**Comment by elylucas** [comment]
Hi @jeffcjohnson,

We have a revamped version of ionic/react-router coming out soon that I think fixes your issue. Could you try installing a dev build of it and let us know?

`npm i @ionic/react@5.3.0-dev.202007062206.b165a26 @ionic/react-router@5.3.0-dev.202007062206.b165a26`

This should be a part of the 5.3 release in the next few weeks.

Thanks

**Comment by jeffcjohnson** [comment]
Hi @elylucas, thanks for the quick response.

With the dev build it works on the first click for `Support` but when I click it a second time it triggers a break point in the React DevTools for Firefox plugin.  When I click the resume button in the debugger, then it gets the max depth error.

```
// --- Welcome to debugging with React DevTools ---
// This debugger statement means that you've enabled the "break on warnings" feature.
// Use the browser's Call Stack panel to step out of this override function-
// to where the original warning or error was logged.
// eslint-disable-next-line no-debugger
debugger;t.apply(void 0,o)};n.__REACT_DEVTOOLS_ORIGINAL_METHOD__=t,// $FlowFixMe property error|warn is not writable.
```

**Comment by elylucas** [comment]
I'm noticing that its only erroring when you go from /login to /support, so it seems to be an issue redirecting back to itself. Do you see the same?

I also noticed that it works fine if you change the Redirect from
  `<Redirect to={{pathname: "/login", state: {from: routeProps.location}}}/>`
to
`<Redirect to="/login" />`

What are you trying to do setting that state?

**Comment by jeffcjohnson** [comment]
Yes, it only errors when going there from /login.

The `from` location is sent in order for the LoginPage to redirect back to the referring page when successfully logged in as demonstrated [here](https://reactrouter.com/web/example/auth-workflow).

```
function LoginPage() {
  let history = useHistory();
  let location = useLocation();

  let { from } = location.state || { from: { pathname: "/" } };
  let login = () => {
    fakeAuth.authenticate(() => {
      history.replace(from);
    });
  };

  return (
    <div>
      <p>You must log in to view the page at {from.pathname}</p>
      <button onClick={login}>Log in</button>
    </div>
  );
}
```

**Comment by elylucas** [comment]
Gotcha, thanks, taking a look into it

**Comment by elylucas** [comment]
Hi @jeffcjohnson,

I noticed passing in the routeProps.location to the state is what is seeming to cause an infinite render somewhere inside react router. This is happening because the value of `location` changes each render. Are you able to pull off just the pathname from location, instead of passing the whole object?:

```
<Redirect to={{pathname: "/login", state: {from: routeProps.location.pathname}}}/>
```

**Comment by jeffcjohnson** [comment]
I tried this in App.tsx which stopped the error:

`<Redirect to={{pathname: "/login", state: {from: {pathname: routeProps.location.pathname}}}}/>`

And this in Login.tsx:

```
interface MyLocationState {
  from: {pathname: string};
}

...

  const location = useLocation<MyLocationState>();
  const { from } = location.state || { from: { pathname: "/" } };
  console.log("from:", from);

  const login = async (e: React.FormEvent) => {
    e.preventDefault();
    setFormSubmitted(true);
    await setIsLoggedIn(true);
    await setUsernameAction(username);
    history.replace(from);
    // history.push('/tabs/schedule', {direction: 'none'});
  };
```

But as seen from the log, the value is overwritten so the redirect won't work:

```
from:  Object { pathname: "/" } Login.tsx:31
from:  Object { pathname: "/support" } Login.tsx:31
from:  Object { pathname: "/login" } Login.tsx:31
```

**Comment by awedev** [comment]
Glad I am not the only one encountered this after being stuck on this for two days. As a workaround, if you wrap Routes with <Switch />, it's unblocked, but you kind of losing transition animation, since each route is mounted exclusively with <Switch />. I noticed the same where passing state is in redirect is actually causing the issue, @elylucas Any updates on this?

@jeffcjohnson did you end up with a solution?

**Comment by liamdebeasi** [comment]
Hi everyone,

Is this still an issue with the latest version of Ionic 6? There have been many improvements to Ionic React since this issue was created, so it's possible that the issue has been resolved.

**Comment by m3thom** [comment]
> Hi everyone,
> 
> Is this still an issue with the latest version of Ionic 6? There have been many improvements to Ionic React since this issue was created, so it's possible that the issue has been resolved.

Yes, I still have this issue with the latest version of Ionic 6 with similar code from the question.

**Comment by liamdebeasi** [comment]
@m3thom Can you provide a GitHub repo I can use to verify this behavior?

**Comment by m3thom** [comment]
> @m3thom Can you provide a GitHub repo I can use to verify this behavior?

Here is the link to the repo. https://github.com/m3thom/maximum-depth-exceeded-reproduce

This repo is based on https://v5.reactrouter.com/web/example/auth-workflow

**Steps to reproduce.**
1. Click 'Tab 3' tab.
2. You will be redirected to the login page.
3. Click 'Log in' button.
4. You will be redirected to the 'Tab 3' page.
5. At 'Tab 3' page. Click 'Sign out' button.
6. Open the console to see the error.

**Comment by m3thom** [comment]
Thank you for your answer.
I fix my code as per your suggestion.

using `routeProps.location.pathname === '/login'` condition
This code works but fails for some scenarios. You would not be able to see the login page even if you are in the /login path

```javascript
return <Route
        {...rest}
        render={(routeProps) => {
            return auth?.user ? (
                children
            ) : (
                routeProps.location.pathname === '/login'
                    ?
                    null
                    :
                    <Redirect
                        to={{
                            pathname: "/login",
                            state: { from: routeProps.location }
                        }}
                    />
            )
        }}
    />
```

Please see the attached video for more information.

https://user-images.githubusercontent.com/39703347/172127811-9b6dd990-07f2-4e7a-b84f-4d39867df0e4.mov

**Comment by sean-perkins** [comment]
Hello @m3thom looking into this issue again. 

I believe the routing structure may be a driving factor to the issue here.

The `/login` page does not fit within mobile tabs navigation rules. The route is a root route, but is being rendered inside of the tabs router outlet, having the tab bar being displayed when the route is active. The tab bar should only be displayed for tab pages, or sub-pages of tabs (that are within the stack).

There are two valid solutions I see with this use case:

### Login is a root route outside of the tabs outlet

Treating the `/login` page as a page outside of the tabs outlet will result in the page being full height (not displaying the tab bar). This pattern is most often use when the entire tabs experience is protected by authentication. 

e.g.:
```jsx
const App: React.FC = () => (
  <ProvideAuth>
    <IonApp>
      <IonReactRouter>
        <IonRouterOutlet>
          <Route exact path="/login">
            <Login />
          </Route>
          <IonTabs>
            <IonRouterOutlet>
              <Route exact path="/tab1">
                <Tab1 />
              </Route>
              <Route exact path="/tab2">
                <Tab2 />
              </Route>

              <PrivateRoute path="/tab3">
                <Tab3 />
              </PrivateRoute>

              <Route exact path="/">
                <Redirect to="/tab1" />
              </Route>
            </IonRouterOutlet>
            <IonTabBar slot="bottom">
              <IonTabButton tab="tab1" href="/tab1">
                <IonIcon icon={triangle} />
                <IonLabel>Tab 1</IonLabel>
              </IonTabButton>
              <IonTabButton tab="tab2" href="/tab2">
                <IonIcon icon={ellipse} />
                <IonLabel>Tab 2</IonLabel>
              </IonTabButton>
              <IonTabButton tab="tab3" href="/tab3">
                <IonIcon icon={square} />
                <IonLabel>Tab 3</IonLabel>
              </IonTabButton>
            </IonTabBar>
          </IonTabs>
        </IonRouterOutlet>
      </IonReactRouter>
    </IonApp>
  </ProvideAuth>
);
```

<video src="https://user-images.githubusercontent.com/13732623/201206944-952dfece-c028-4397-9cdb-fa9455cf33e2.mp4"></video>

I did have to tweak some of the logic around the location state, repro: https://github.com/sean-perkins/ionic-gh-21717

### Use a modal in place of a protected route

If the true desire is to allow unauthenticated users to access tab 1 and tab 2, but have to sign into access tab 3, a modal experience that takes over the fullscreen when tab 3 is activated would be desired. 

I experimented with trying to render the private route inside of the tab3 routing structure, but this leads to other problems, that activating the tab bar button for tab3 attempts to route to `/tab3`, but is redirected to `/tab3/login` when unauthenticated. This results in a push transition which looks bad and is probably an anti-pattern for navigation. 

Can you try either of these suggestions and let me know if that resolves your issue or if you observe anything else?

**Comment by m3thom** [comment]
Your suggestions work for me. Thank you.

**Comment by sean-perkins** [comment]
Awesome, I am going to close this issue out as completed, as the main issue reported is resolved. 

If any run into any other issues, feel free to open a new issue and the team would be happy to take a look.
